module/VuFind/src/VuFind/Crypt/BlockCipher.php

```diff
@@ -0,0 +1,550 @@
+     */
+    protected $openSslKey = null;
+
+    /**
+     * Constructor
+     *
+     * @param array $options Options (supported key: algorithm)
+     */
+    public function __construct(protected array $options = [])
+    {
+        if (!extension_loaded('openssl')) {
+            throw new \RuntimeException('OpenSSL extension required!');
+        }
+        if (isset($options['algorithm'])) {
+            $this->setAlgorithm($options['algorithm']);
+        }
+    }
+
+    /**
+     * Pad the string using PKCS#7
+     *
+     * @param string $string    String to pad
+     * @param int    $blockSize Target size
+     *
+     * @return string
+     */
+    protected function pkcs7Pad(string $string, int $blockSize = 32): string
+    {
+        $pad = $blockSize - (mb_strlen($string, '8bit') % $blockSize);
+        return $string . str_repeat(chr($pad), $pad);
+    }
+
+    /**
+     * Unpad a PKCS#7 string
+     *
+     * @param string $string String to unpad
+     *
+     * @return string
+     */
+    protected function pkcs7Strip(string $string): string
+    {
+        $end = mb_substr($string, -1, null, '8bit');
+        $last = ord($end);
+        $len = mb_strlen($string, '8bit') - $last;
+        if (mb_substr($string, $len, null, '8bit') === str_repeat($end, $last)) {
+            return mb_substr($string, 0, $len, '8bit');
+        }
+        throw new \Exception('Cannot unpad string');
+    }
+
+    /**
+     * Look up appropriate key size for selected algorithm.
+     *
+     * @return int
+     */
+    protected function getKeySize(): int
+    {
+        return $this->keySizes[$this->algo];
+    }
+
+    /**
+     * Set the encryption key for use by OpenSSL; note that this may be truncated by getOpenSslKey()
+     * if it exceeds the length limit for the selected algorithm.
+     *
+     * @param string $key OpenSSL encryption key
+     *
+     * @return static
+     * @throws InvalidArgumentException
+     */
+    protected function setOpenSslKey(string $key): static
+    {
+        $keyLen = mb_strlen($key, '8bit');
+
+        if ($keyLen < $this->getKeySize()) {
+            throw new InvalidArgumentException('OpenSSL key is too short.');
+        }
+
+        $this->openSslKey = $key;
+        return $this;
+    }
+
+    /**
+     * Get the key to use for OpenSSL encryption/decryption.
+     *
+     * @return string
+     */
+    protected function getOpenSslKey(): string
+    {
+        if (empty($this->openSslKey)) {
+            throw new \Exception('OpenSSL key not set');
+        }
+        return mb_substr($this->openSslKey, 0, $this->getKeySize(), '8bit');
+    }
+
+    /**
+     * Set the encryption algorithm (cipher)
+     *
+     * @param string $algo New algorithm
+     *
+     * @return static
+     * @throws InvalidArgumentException
+     */
+    protected function setAlgorithm(string $algo): static
+    {
+        if (!in_array($this->encryptionAlgos[$algo] . '-cbc', openssl_get_cipher_methods(true))) {
```

Comment:
The original Laminas code had a hard-coded cbc (possibly as a "least common denominator"), but I've revised our version to take mode into consideration via a new setAlgorithmAndMode() method that gets wrapped by setAlgorithm(). I think this is safe for now, and it may make it easier to introduce new options in future if the need arises.